windows/src/desktop/kmshell/main/Keyman.System.UpdateStateMachine.pas

```diff
@@ -807,14 +819,38 @@ procedure DownloadingState.HandleCheck;
 end;
 
 function DownloadingState.HandleKmShell;
+var
+  FMutex: TKeymanMutex;
 begin
-  // Downloading state, in other process, so continue
+  // Check to ensure a download process is running if not
+  // clean up return to the the idle state and check for updates
+  FMutex := TKeymanMutex.Create('KeymanDownloading');
+  if FMutex.MutexOwned then
+  begin
+    bucStateContext.RemoveCachedFiles;
+    FreeAndNil(FMutex);
+    ChangeState(IdleState);
+    bucStateContext.CurrentState.HandleCheck;
+  end;
+  FreeAndNil(FMutex);
```

Comment:
Does it matter that we call `FreeAndNil` twice if `FMutex.MutexOwned`? (same below)

Reply:
Yeah good catch, I should exit early to avoid the redundant call. I have also added a comment to make clear to future me or other developers that the mutex needs to be freed before changing state and triggering a force check. Updated to use a try .. finally handling.

Reply:
FWIW, `FreeAndNil` is safe to call on an already-nil variable, but it's definitely a code smell in any case :grin:

Reply:
> Updated to use a try .. finally handling.

Note that exiting early still runs the finally section